talk/basicconcepts/headersinterfaces.tex

```diff
@@ -73,3 +73,51 @@
     \end{cppcode*}
   \end{block}
 \end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Header / source separation}
+  \begin{goodpractice}[Header / source]{Headers and source files}
+    \begin{itemize}
+      \item Headers should only contain declarations that are necessary to know the interface of functions / classes
+      \item Headers might be included and compiled many times, so it is good to keep them short
+      \begin{itemize}
+        \item The same goes for \cppinline{#include} statements in headers
+      \end{itemize}
+      \item Implementation files only get compiled once, so put necessary includes and most of the code here
+        \begin{itemize}
+          \item Exception: Short functions can be in headers
+          \item Templates (see later) should be in headers
+        \end{itemize}
```

Comment:
Implementations of `constexpr` functions must be in headers. Maybe add that.

Also: every function where inlining would be a benefit, or where the function could be auto vectorized, is a good candidate for being in a header. But that is a big discussion with trade-offs and personal preferences, so hard to give a guideline here.

Reply:
`inline` keyword comes later, and only in the advanced course. Vectorization is not even mentioned in this course I believe, so I would skip this comment here.

Reply:
Sure, we would not discuss vectorization. I am just slighly afraid that the students would start to split *every* function between header and source and needlessly pessimize their code. But maybe I just have different preferences here. Also link-time-optimization (LTO) gets rid of the problem.